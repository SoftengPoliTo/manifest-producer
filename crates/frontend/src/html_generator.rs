use std::{collections::HashMap, fs::File, io::Write};

use crate::{
    error::Result,
    subtrees_generator::{build_subtrees, identify_subtrees},
    tree_generator::build_tree,
    TreeNode,
};
use manifest_producer_backend::{BasicInfo, FunctionNode};

use minijinja::{context, Environment};
use serde_json;

/// Generates HTML reports from analysis results.
///
/// This function orchestrates the creation of various HTML pages that summarize the results
/// of the analysis, including:
///
/// - An index page with general information.
/// - A page listing detected functions.
/// - A root page displaying entry points (root nodes).
///
/// Additionally, it identifies subtrees, cleans redundant nodes, and constructs trees
/// for visual representation.
///
/// # Arguments
///
/// - `basic_info`: Metadata about the binary being analyzed.
/// - `detected_functions`: A map of function names to their associated `FunctionNode` objects.
/// - `root_nodes`: A vector of root function names identified during analysis.
/// - `output_path`: The directory where the HTML files should be saved.
///
/// # Workflow
///
/// 1. Renders the main HTML pages using helper functions.
/// 2. Iterates over `root_nodes` to build and render subtrees and visual trees.
/// 3. Cleans temporary structures (`node_roots` and `sub_trees`) after processing each root.
///
/// # Returns
///
/// - `Ok(())`: If all HTML files are generated successfully.
/// - `Err(e)`: If any operation fails, an error is returned.
///
/// # Errors
///
/// Errors can arise from:
/// - File I/O failures during HTML generation.
/// - Invalid or incomplete data structures passed to the function.
/// - Errors in subtree or tree construction.
///
/// # Example
///
/// ```
/// use manifest_producer_frontend::html_generator;
/// use manifest_producer_backend::BasicInfo;
/// use std::collections::HashMap;
/// use manifest_producer_frontend::html_generator::html_generator;
///
/// let basic_info = BasicInfo::new("example.elf", "Executable");
/// let detected_functions = HashMap::new();
/// let root_nodes = vec!["main".to_string()];
/// let output_path = "./output";
///
/// if let Err(err) = html_generator(&basic_info, &detected_functions, &root_nodes, output_path) {
///     eprintln!("HTML generation failed: {}", err);
/// }
/// ```
#[allow(clippy::implicit_hasher)]
pub fn html_generator(
    basic_info: &BasicInfo,
    detected_functions: &HashMap<String, FunctionNode>,
    root_nodes: &Vec<String>,
    output_path: &str,
) -> Result<()> {

    #[cfg(feature = "progress_bar")]
    let pb = {
        let pb = indicatif::ProgressBar::new((root_nodes.len()*2) as u64);
        pb.set_message("Building call graphs:".to_string());
        pb.set_style(
            indicatif::ProgressStyle::default_bar()
                .template("{msg}\n{wide_bar} {pos}/{len} [{elapsed_precise}]")?,
        );
        pb
    };

    let mut node_roots: HashMap<String, FunctionNode> = HashMap::new();
    let mut sub_trees: HashMap<String, TreeNode> = HashMap::new();
    let mut id_counter = 0;

    render_index_page(
        basic_info,
        detected_functions.len(),
        root_nodes.len(),
        output_path,
    )?;
    render_functions_page(detected_functions, output_path)?;
    render_root_page(root_nodes, output_path)?;

    for root in root_nodes {
        #[cfg(feature = "progress_bar")]
        pb.inc(1);

        // Step 1: Identification of subtrees
        identify_subtrees(root, detected_functions, &mut node_roots);

        // Step 2: Cleaning nodes with jmp equal to zero and creating subtrees
        build_subtrees(
            &mut node_roots,
            detected_functions,
            &mut sub_trees,
            &mut id_counter,
        );

        // Step 3: Construction of the tree
        build_tree(
            root,
            detected_functions,
            &mut sub_trees,
            &mut id_counter,
            output_path,
        )?;

        // Empty the structures for the next cycle
        node_roots.clear();
        sub_trees.clear();

        #[cfg(feature = "progress_bar")]
        pb.inc(1);
    }
    #[cfg(feature = "progress_bar")]
    pb.finish_with_message("Call graphs built!");
    Ok(())
}

pub(crate) fn render_index_page(
    basic_info: &BasicInfo,
    num_func: usize,
    num_root: usize,
    output_path: &str,
) -> Result<()> {
    let mut env = Environment::new();
    env.add_template("index.html", include_str!("templates/index.html"))?;

    let template = env.get_template("index.html")?;
    let rendered = template.render(context! {
        basic_info => basic_info,
        num_func => num_func,
        num_root => num_root
    })?;

    let mut file = File::create(format!("{output_path}/index.html"))?;
    file.write_all(rendered.as_bytes())?;
    Ok(())
}

pub(crate) fn render_functions_page(
    detected_functions: &HashMap<String, FunctionNode>,
    output_path: &str,
) -> Result<()> {
    let functions: Vec<FunctionNode> = detected_functions.values().cloned().collect();

    let mut env = Environment::new();
    env.add_template(
        "functions_list.html",
        include_str!("templates/functions_list.html"),
    )?;
    let template = env.get_template("functions_list.html")?;
    let rendered = template.render(context! {
        functions => functions
    })?;

    let mut file = File::create(format!("{output_path}/functions_list.html"))?;
    file.write_all(rendered.as_bytes())?;

    Ok(())
}

pub(crate) fn render_root_page(roots: &Vec<String>, output_path: &str) -> Result<()> {
    let mut env = Environment::new();
    env.add_template(
        "root_functions.html",
        include_str!("templates/root_functions.html"),
    )?;

    let template = env.get_template("root_functions.html")?;
    let rendered = template.render(context! {
        roots => roots,
    })?;

    let mut file = File::create(format!("{output_path}/root_functions.html"))?;
    file.write_all(rendered.as_bytes())?;
    Ok(())
}

pub(crate) fn render_tree_page(
    root_name: &str,
    js_tree: &TreeNode,
    output_path: &str,
) -> Result<()> {
    let mut env = Environment::new();
    env.add_template("call_tree.html", include_str!("templates/call_tree.html"))?;

    let template = env.get_template("call_tree.html")?;

    let js_tree_json = serde_json::to_string(&js_tree)?;

    let rendered = template.render(context! {
        root_name => root_name,
        js_tree => js_tree_json,
    })?;

    let mut file = File::create(format!("{output_path}/call_trees/{root_name}.html"))?;
    file.write_all(rendered.as_bytes())?;

    Ok(())
}
