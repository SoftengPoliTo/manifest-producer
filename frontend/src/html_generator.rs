use std::{collections::HashMap, fs::File, io::Write};

use common::{
    error::Result,
    minijinja::{context, Environment},
    open, serde_json, BasicInfo, CallTree, TreeNode, FUNC,
};

pub fn html_generator(
    basic_info: BasicInfo,
    num_func: usize,
    num_root: usize,
    functions: &Vec<FUNC>,
    forest: &HashMap<String, CallTree>,
    disassembly: &Vec<(String, String)>,
    roots: &Vec<String>,
) -> Result<()> {
    render_index_page(basic_info, num_func, num_root)?;
    render_functions_page(functions, forest)?;
    render_disassembly_page(disassembly)?;
    render_root_page(roots)?;

    Ok(())
}

pub fn open_index_page() -> Result<()> {
    let index_path = "./public/index.html";
    open::that(index_path)?;
    Ok(())
}

pub fn render_index_page(basic_info: BasicInfo, num_func: usize, num_root: usize) -> Result<()> {
    let mut env = Environment::new();
    env.add_template("index.html", include_str!("templates/index.html"))?;

    let template = env.get_template("index.html")?;
    let rendered = template.render(context! {
        basic_info => basic_info,
        num_func => num_func,
        num_root => num_root
    })?;

    let mut file = File::create("public/index.html")?;
    file.write_all(rendered.as_bytes())?;
    Ok(())
}

pub fn render_functions_page(
    functions: &Vec<FUNC>,
    forest: &HashMap<String, CallTree>,
) -> Result<()> {
    let mut combined: Vec<(FUNC, CallTree)> = Vec::new();

    for func in functions {
        let call_tree = forest.get(&func.name).cloned();
        if let Some(call_tree) = call_tree {
            combined.push((func.clone(), call_tree));
        }
    }

    let mut env = Environment::new();
    env.add_template(
        "functions_list.html",
        include_str!("templates/functions_list.html"),
    )?;
    let template = env.get_template("functions_list.html")?;
    let rendered = template.render(context! {
        combined => combined
    })?;

    let mut file = File::create("public/functions_list.html")?;
    file.write_all(rendered.as_bytes())?;

    Ok(())
}

pub fn render_disassembly_page(disassembly: &Vec<(String, String)>) -> Result<()> {
    let mut env = Environment::new();
    env.add_template(
        "disassembly_view.html",
        include_str!("templates/disassembly_view.html"),
    )?;

    let template = env.get_template("disassembly_view.html")?;
    let rendered = template.render(context! {
        disassembly => disassembly,
    })?;

    let mut file = File::create("public/disassembly_view.html")?;
    file.write_all(rendered.as_bytes())?;
    Ok(())
}

pub fn render_root_page(roots: &Vec<String>) -> Result<()> {
    let mut env = Environment::new();
    env.add_template(
        "root_functions.html",
        include_str!("templates/root_functions.html"),
    )?;

    let template = env.get_template("root_functions.html")?;
    let rendered = template.render(context! {
        roots => roots,
    })?;

    let mut file = File::create("public/root_functions.html")?;
    file.write_all(rendered.as_bytes())?;
    Ok(())
}

pub fn render_tree_page(root_name: &str, js_tree: &TreeNode) -> Result<()> {
    let mut env = Environment::new();
    env.add_template("call_graph.html", include_str!("templates/call_graph.html"))?;

    let template = env.get_template("call_graph.html")?;

    let js_tree_json = serde_json::to_string(&js_tree)?;

    let rendered = template.render(context! {
        root_name => root_name,
        js_tree => js_tree_json,
    })?;

    let mut file = File::create(format!("public/call_graphs/{}.html", root_name))?;
    file.write_all(rendered.as_bytes())?;

    Ok(())
}
