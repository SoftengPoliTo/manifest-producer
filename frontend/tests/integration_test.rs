use std::{collections::HashMap, env::temp_dir, fs, path::Path};

use manifest_producer_backend::{BasicInfo, FunctionNode};
use manifest_producer_frontend::html_generator::html_generator;
use walkdir::WalkDir;

#[test]
fn run_frontend_test() {
    let info = BasicInfo::new("Test-Frontend", "Executable")
        .file_size(916432)
        .arch("x86_64")
        .pie(false)
        .stripped(false)
        .static_linking("Statically linked")
        .language("Rust".to_string())
        .entry_point(4199776);

    let mut node1 = FunctionNode::new("root".to_string(), 4499220, 4499459);
    let mut node2 = FunctionNode::new("child1".to_string(), 4445450, 4445459);
    let mut node3 = FunctionNode::new("child2".to_string(), 4445460, 4445469);
    let mut node21 = FunctionNode::new("child11".to_string(), 4499240, 4499249);
    let mut node22 = FunctionNode::new("child12".to_string(), 4445480, 4445489);
    let mut node31 = FunctionNode::new("child21".to_string(), 4455460, 4455469);

    node1.children.push(node2.name.clone());
    node1.children.push(node3.name.clone());
    node2.children.push(node21.name.clone());
    node2.children.push(node22.name.clone());
    node3.children.push(node31.name.clone());

    let disassembly1 = "0x494340:\tendbr64\t\n0x494344:\tmov\t8(%rdx), %rcx\n0x494348:\tmov\t$1, %eax\n0x49434d:\tcmp\t%rcx, 8(%rsi)\n0x494351:\tja\t0x494355\n0x494353:\tsbb\t%eax, %eax\n0x494355:\tret\t\n".to_string();
    let disassembly2 = "0x44ff50:\tendbr64\t\n0x44ff54:\tmov\t$0xa, %eax\n0x44ff59:\tsyscall\t\n0x44ff5b:\tcmp\t$-0xfff, %rax\n0x44ff61:\tjae\t0x44ff64\n0x44ff63:\tret\t\n0x44ff64:\tmov\t$18446744073709551544, %rcx\n0x44ff6b:\tneg\t%eax\n0x44ff6d:\tmov\t%eax, %fs:(%rcx)\n0x44ff70:\tor\t$0xffffffffffffffff, %rax\n0x44ff74:\tret\t\n".to_string();
    let disassembly3= "0x4368c0:\tendbr64\t\n0x4368c4:\tmov\t$18446744073709551528, %rax\n0x4368cb:\tmov\t%fs:(%rax), %rdx\n0x4368cf:\tnop\t(%rax, %rax)\n".to_string();
    let disassembly21= "0x4368c0:\tendbr64\t\n0x4368c4:\tmov\t$18446744073709551528, %rax\n0x4368cb:\tmov\t%fs:(%rax), %rdx\n0x4368cf:\tnop\t(%rax, %rax)\n".to_string();
    let disassembly22= "0x4368c0:\tendbr64\t\n0x4368c4:\tmov\t$18446744073709551528, %rax\n0x4368cb:\tmov\t%fs:(%rax), %rdx\n0x4368cf:\tnop\t(%rax, %rax)\n".to_string();
    let disassembly31= "0x4368c0:\tendbr64\t\n0x4368c4:\tmov\t$18446744073709551528, %rax\n0x4368cb:\tmov\t%fs:(%rax), %rdx\n0x4368cf:\tnop\t(%rax, %rax)\n".to_string();

    node1.set_disassembly(disassembly1);
    node2.set_disassembly(disassembly2);
    node3.set_disassembly(disassembly3);
    node21.set_disassembly(disassembly21);
    node22.set_disassembly(disassembly22);
    node31.set_disassembly(disassembly31);

    let mut detected_functions: HashMap<String, FunctionNode> = Default::default();
    detected_functions.insert(node1.name.clone(), node1);
    detected_functions.insert(node2.name.clone(), node2);
    detected_functions.insert(node3.name.clone(), node3);
    detected_functions.insert(node21.name.clone(), node21);
    detected_functions.insert(node22.name.clone(), node22);
    detected_functions.insert(node31.name.clone(), node31);

    let root_nodes = vec!["root".to_string()];

    let output_path = setup_test_environment();

    html_generator(info, &detected_functions, &root_nodes, &output_path).unwrap();

    let snapshot_path = Path::new("./snapshots");
    compare_generated_html(snapshot_path, Path::new(&output_path));
}

fn setup_test_environment() -> String {
    let test_dir = temp_dir().join("frontend");
    let tree_dir = test_dir.join("call_trees");
    let json_dir = test_dir.join("json");

    for dir in [&test_dir, &tree_dir, &json_dir] {
        if !dir.exists() {
            fs::create_dir_all(dir).unwrap();
        }
    }

    test_dir.to_str().unwrap().to_owned()
}

fn compare_generated_html(snapshot_path: &Path, output_path: &Path) {
    for entry in WalkDir::new(output_path).into_iter() {
        entry.map_or((), |e| {
            if e.path().is_file() && !e.path().components().any(|c| c.as_os_str() == "call_trees") {
                compare(snapshot_path, output_path, e.path());
            }
        })
    }
}

fn compare(snapshot_path: &Path, output_path: &Path, entry: &Path) {
    let content = fs::read_to_string(entry).unwrap();
    let name = entry.file_name().and_then(|v| v.to_str()).unwrap();

    insta::with_settings!({
        snapshot_path => snapshot_path
        .join(entry.strip_prefix(output_path).unwrap())
        .parent()
        .unwrap(),
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, content);
    });
}
