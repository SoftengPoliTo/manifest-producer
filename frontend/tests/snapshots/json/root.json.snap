---
source: frontend/tests/integration_test.rs
expression: content
snapshot_kind: text
---
{
  "id": 0,
  "text": "root",
  "children": [
    {
      "id": 1,
      "text": "child1",
      "children": [
        {
          "id": 2,
          "text": "child11",
          "children": null
        },
        {
          "id": 3,
          "text": "child12",
          "children": null
        }
      ]
    },
    {
      "id": 4,
      "text": "child2",
      "children": [
        {
          "id": 5,
          "text": "child21",
          "children": null
        }
      ]
    }
  ]
}
